### ISSUE TITLE
Followup to PR #1 ProNE/FastRP Embeddings

### ISSUE THREAD

**Issue by ghost** [Refefer/cloverleaf#7]
## TLDR 

Take [fastRP](https://github.com/shur-complement/cloverleaf/blob/ProNE/examples/dataset/pinterest/fast_cloverleaf.py) and [ProNE](https://github.com/shur-complement/cloverleaf/blob/ProNE/examples/dataset/pinterest/prone_cloverleaf.py)  embeddings for a whirl, followed by `./eval.sh whatever-cloverleaf.embeddings`. In testing, I am seeing:

| method | metric | score   |
|-------------|-----------|-----------|
| ProNE| HR @ 5 | 0.6972302769723028 |
| FastRP | HR @ 5 | 0.6706329367063294 |
| FastRP+ProNE | HR @ 5 | 0.7018596280743852 |

| method | metric | score   |
|-------------|-----------|-----------|
| ProNE| HR @ 10 | 0.8535146485351465|
| FastRP | HR @ 10 | 0.8425157484251575 |
| FastRP+ProNE | HR @ 10 | 0.8632136786321368 |

Rust implementations at: 

1) https://github.com/shur-complement/cloverleaf/blob/ProNE/src/algos/fast_random_projection.rs
2) https://github.com/shur-complement/cloverleaf/blob/ProNE/src/algos/prone.rs

## Overview

I went ahead and implemented ProNE and FastRP in Rust for Cloverleaf, and included reference implementations in Python. `FastRP` is indeed blazingly fast, and combined with the "spectral propagation" step from ProNE seems to give the best of both. The authors of FastRP seem to be right, in that the factorization method is less important than the choice of matrix to factorize. Sparse projection is _much_ cheaper than randomized SVD. Some work is left to tidy all this up into a PR if interested.

## Caveats

All of this is built on sparse matrix math, which I included via nalgebra-sparse. However, since that is a distinct type from Cloverleaf's CSR, I did some plumbing to copy memory. Something to consider is either adopting a matrix library and using their CSR type, or hand-rolling some matrix ops into cloverleaf.

Separately, there was an issue with Pinterest data with duplicate edges, and it appears that no edge-deduping occurs in cloverleaf, so we had duplicate indices in `CSR` at runtime (and they were not sorted?). I deduped the graph manually, and added index sorting. We should probably do that at `CSR` construction time.

There is also some efficiency left on the table on the Rust side, we can reuse allocations, which would help for scaling to _very_ large graphs, but I have not yet taken the time to do this.
